node/network/dispute-distribution/src/receiver/batches/waiting_queue.rs

```diff
@@ -0,0 +1,204 @@
+
+use std::{cmp::Ordering, collections::BinaryHeap, time::Instant};
+
+use futures::future::pending;
+use futures_timer::Delay;
+
+/// Wait asynchronously for given `Instant`s one after the other.
+///
+/// `PendingWake`s can be inserted and `WaitingQueue` makes `wait_ready()` to always wait for the
+/// next `Instant` in the queue.
+pub struct WaitingQueue<Payload> {
+	/// All pending wakes we are supposed to wait on in order.
+	pending_wakes: BinaryHeap<PendingWake<Payload>>,
+	/// Wait for next `PendingWake`.
+	waker: Option<Delay>,
+}
+
+/// Represents some event waiting to be processed at `ready_at`.
+///
+/// This is an event in `WaitingQueue`. It provides an `Ord` instance, that sorts descending with
+/// regard to `Instant` (so we get a `min-heap` with the earliest `Instant` at the top).
+#[derive(Eq, PartialEq)]
+pub struct PendingWake<Payload> {
+	pub payload: Payload,
+	pub ready_at: Instant,
+}
+
+impl<Payload: Eq + Ord> WaitingQueue<Payload> {
+	/// Get a new empty `WaitingQueue`.
+	///
+	/// If you call `pop` on this queue immediately, it will always return `Poll::Pending`.
+	pub fn new() -> Self {
+		Self { pending_wakes: BinaryHeap::new(), waker: None }
+	}
+
+	/// Push a `PendingWake`.
+	///
+	/// The next call to `wait_ready` will make sure to wake soon enough to process that new event in a
+	/// timely manner.
+	pub fn push(&mut self, wake: PendingWake<Payload>) {
+		self.pending_wakes.push(wake);
+		// Reset waker as it is potentially obsolete now:
+		self.waker = None;
+	}
+
+	/// Pop the next ready item.
+	///
+	/// This function does not wait, if nothing is ready right now as determined by the passed
+	/// `now` time stamp, this function simply returns `None`.
+	pub fn pop_ready(&mut self, now: Instant) -> Option<PendingWake<Payload>> {
+		let is_ready = self.pending_wakes.peek().map_or(false, |p| p.ready_at <= now);
+		if is_ready {
+			Some(self.pending_wakes.pop().expect("We just peeked. qed."))
+		} else {
+			None
+		}
+	}
+
+	/// Don't pop, just wait until something is ready.
+	///
+	/// Once this function returns `Poll::Ready(())` `pop_ready()` will return `Some`, if passed
+	/// the same `Instant`.
+	///
+	/// Whether ready or not is determined based on the passed time stamp `now` which should be the
+	/// current time as returned by `Instant::now()`
+	///
+	/// This function waits asynchronously for an item to become ready. If there is no more item,
+	/// this call will wait forever (return Poll::Pending without scheduling a wake).
+	pub async fn wait_ready(&mut self, now: Instant) {
+		if let Some(waker) = &mut self.waker {
+			// Previous timer was not done yet.
+			waker.await
+		}
+
+		let next_waiting = self.pending_wakes.peek();
+		let is_ready = next_waiting.map_or(false, |p| p.ready_at <= now);
+		if is_ready {
+			return
+		}
+
+		self.waker = next_waiting.map(|p| Delay::new(p.ready_at.duration_since(now)));
+		match &mut self.waker {
+			None => return pending().await,
```

Comment:
Not a stupid question at all. On it's own it would exactly do that, but that would still be logically valid - "you are waiting on an empty queue, what did you expect?" The thing is calling code will either check whether it is empty or use that function in a `select` where it also waits for other stuff to happen. And here it becomes very clear why this is the correct behavior - we absolutely don't want to wake here, as there is nothing to wake on - we can move on to other futures. Returning `Ready` would result in a busy loop.

Reply:
I see. The code in `receive_message` (node/network/dispute-distribution/src/receiver/mod.rs) calls it but actually doesn't awaits it - only polls it won't block.